[Data][LLM] Support vLLM structured outputs + backward-compat for guided_decoding…

Sathyanarayanaa-T opened the pull request:
### **Description**
This PR adds backward-compatible support for vLLM's migration from the deprecated `guided_decoding` parameter to the new `structured_outputs` API (vLLM v0.9+).

**Key changes:**

- `vllm_engine_stage.py`:

   - Added support for the new `structured_outputs` parameter in `sampling_params`
   - Added backward compatibility for legacy `guided_decoding` with a deprecation warning (logged once per worker)
   - When both parameters are present, `structured_outputs` takes precedence silently
   - Added TODO comment for `guided_decoding` removal in Ray 2.56.0

- `test_vllm_engine_stage.py`:
  - Combined duplicate JSON tests into a single parametrized test covering both `guided_decoding` and `structured_outputs` paths

- Documentation updates:

  - Updated `guided_decoding` → `structured_outputs` in examples:
    - LLM batch inference: 
      - [vllm-with-structural-output.ipynb](ray/doc/source/llm/examples/batch/vllm-with-structural-output.ipynb)
    - E2E audio example: 
      - [e2e-audio/README.md](ray/doc/source/ray-overview/examples/e2e-audio/README.md)
      - [e2e-audio/README.ipynb](ray/doc/source/ray-overview/examples/e2e-audio/README.ipynb)
    - Entity recognition example: 
      - [entity-recognition-with-llms/README.md](ray/doc/source/ray-overview/examples/entity-recognition-with-llms/README.md)
      - [entity-recognition-with-llms/README.ipynb](ray/doc/source/ray-overview/examples/entity-recognition-with-llms/README.ipynb)

**Why**
Users relying on the old `guided_decoding` API will continue to work without code changes, while new code should adopt the recommended `structured_outputs` parameter.

### **Related issues**
Fixes #59321

### **Additional information**
Minimum vLLM version: v0.9.0

**Migration example:**

Old (still works):

```python
sampling_params = {'guided_decoding': {'json': <json_schema>}}
```
New (recommended):

```python
  sampling_params = {"structured_outputs": {"json": <json_schema>}}
```
**Testing:**

```bash
python -m pytest python/ray/llm/tests/batch/gpu/stages/test_vllm_engine_stage.py::test_vllm_wrapper_json -v
```



vLLM docs: https://docs.vllm.ai/en/stable/features/structured_outputs/

jeffreywang-anyscale commented on README.rst:
```diff
@@ -75,6 +75,42 @@ More Information
+backward-compatible mapping for `guided_decoding`, but you should migrate to the
+new `response_format`/structured outputs pattern. See the implementation notes:
+
+- `ISSUE_59321 implementation summary <./ISSUE_59321_IMPLEMENTATION_SUMMARY.md>`__
```
Could you point me to what file this references?

jeffreywang88 reviewed (changes requested):
Hey @Sathyanarayanaa-T thanks for your contribution! Couple thoughts:
- Your intention for deprecating the `guided_decoding` field in `sampling_params` to align with vLLM is great, but it seems like the logic to parse `structured_outputs` from `sampling_params` is missing.
- If you're just trying to add deprecation warnings to the `guided_decoding` field in Ray Data LLM, I don't think that's necessary.
  - Without the new implementation, we won't deprecate the existing `guided_decoding` field.
  - Even though we have switched to vLLM's structured output API in Ray Data LLM's internal implementation, the user-facing interface remains unchanged. That is, users can continue to specify `guided_decoding`.
- You mentioned that the changes were tested with `test_vllm_wrapper_json_new_api`. I don't find the test in the commits. Could you please point me to the tests and elaborate on the behavior of the new API?

Sathyanarayanaa-T commented on python/ray/llm/_internal/batch/stages/vllm_engine_stage.py:
```diff
@@ -363,11 +363,20 @@ async def _prepare_llm_request(self, row: Dict[str, Any]) -> vLLMEngineRequest:
+                # Log deprecation warning for the old guided_decoding parameter
+                logger.warning(
+                    "The 'guided_decoding' parameter is deprecated. "
+                    "Please use 'structured_output_config' in engine initialization "
```
You're right, I realize now I pointed to a parameter that isn't actually implemented yet. I'll update the engine logic to properly accept structured_output and then fix this warning message so it points to the right place. My bad

Sathyanarayanaa-T committed:
Add structured_output support and deprecate guided_decoding (#59321)

- Add support for new 'structured_output' field in sampling_params
- Deprecate legacy 'guided_decoding' with backward compatibility
- Add test_vllm_wrapper_json_new_api to verify new API
- Fix deprecation warning to point to correct parameter
- Remove migration notes from README.rst (wrong location)

Signed-off-by: Sathyanarayanaa-T <tsathyanarayanaa@gmail.com>

jeffreywang-anyscale commented on python/ray/llm/_internal/batch/stages/vllm_engine_stage.py:
```diff
@@ -363,11 +365,31 @@ async def _prepare_llm_request(self, row: Dict[str, Any]) -> vLLMEngineRequest:
             sampling_params = row.pop("sampling_params")
-            if "guided_decoding" in sampling_params:
+            # Handle new structured_output parameter (preferred)
+            if "structured_output" in sampling_params:
```
Let's use `structured_outputs` (plural) to align with vLLM's [naming](https://github.com/vllm-project/vllm/blob/04147dcfa70fb7228ce9e2f88fa7dd41631d17f0/vllm/sampling_params.py#L227).

Sathyanarayanaa-T replied:
Updated to `structured_outputs` to match vLLM.

jeffreywang-anyscale commented on python/ray/llm/tests/batch/gpu/stages/test_vllm_engine_stage.py:
```diff
@@ -440,6 +443,61 @@ class AnswerModel(BaseModel):
     wrapper.shutdown()
 
 
+@pytest.mark.asyncio
```
Could we use `pytest.mark.parametrize` to pass `"guided_decoding"` or `"structured_outputs"` into `sampling_params`? This helps avoid duplicated code as test_vllm_wrapper_json and test_vllm_wrapper_json_new_api are now mostly identical.

jeffreywang-anyscale commented on python/ray/llm/_internal/batch/stages/vllm_engine_stage.py:
```diff
@@ -363,11 +363,20 @@ async def _prepare_llm_request(self, row: Dict[str, Any]) -> vLLMEngineRequest:
 
         if self.task_type == vLLMTaskType.GENERATE:
             sampling_params = row.pop("sampling_params")
+            # Handle legacy guided_decoding parameter for backward compatibility
```
Just realized that vLLM uses `structured_outputs` (plural). Let's follow vLLM's naming.

Sathyanarayanaa-T committed:
Update python/ray/llm/_internal/batch/stages/vllm_engine_stage.py

Co-authored-by: Jeffrey Wang <jeffreywang@anyscale.com>
Signed-off-by: Sathyanarayanaa <151912659+Sathyanarayanaa-T@users.noreply.github.com>
Signed-off-by: Sathyanarayanaa-T <tsathyanarayanaa@gmail.com>

jeffreywang-anyscale commented:
/gemini review

Sathyanarayanaa-T commented:
Hi @jeffreywang-anyscale,
I've integrated the changes suggested by the Gemini review and pushed the updates. It should be good to go whenever you have a chance to take a look.

kouroshHakha reviewed (approved):
LG
